va/http.go

```diff
@@ -0,0 +1,488 @@
+	}
+
+	// Build a URL with the target's IP address and port
+	targetIP := target.ip()
+	if targetIP == nil {
+		return nil, record, fmt.Errorf(
+			"host %q has no IP addresses remaining to use",
+			target.host)
+	}
+	record.AddressUsed = targetIP
+	url := httpValidationURL(targetIP, target.path, target.port)
+	record.URL = url.String()
+
+	// If there's no provided HTTP request to mutate (e.g. a redirect request
+	// we're following as part of a validation) then construct a new initial HTTP
+	// GET request for the validation.
+	if req == nil {
+		var err error
+		req, err = http.NewRequest("GET", url.String(), nil)
+		if err != nil {
+			return nil, record, err
+		}
+		// Immediately reconstruct the request using the validation context
+		req = req.WithContext(ctx)
+		if va.userAgent != "" {
+			req.Header.Set("User-Agent", va.userAgent)
+		}
+		// Some of our users use mod_security. Mod_security sees a lack of Accept
+		// headers as bot behavior and rejects requests. While this is a bug in
+		// mod_security's rules (given that the HTTP specs disagree with that
+		// requirement), we add the Accept header now in order to fix our
+		// mod_security users' mysterious breakages. See
+		// <https://github.com/SpiderLabs/owasp-modsecurity-crs/issues/265> and
+		// <https://github.com/letsencrypt/boulder/issues/1019>. This was done
+		// because it's a one-line fix with no downside. We're not likely to want to
+		// do many more things to satisfy misunderstandings around HTTP.
+		req.Header.Set("Accept", "*/*")
+	}
+
+	// Override the request's target URL and Host
+	req.URL = url
+	req.Host = target.host
+	return req, record, nil
+}
+
+// fetchHTTPSimple invokes processHTTPValidation and if an error result is
+// returned, converts it to a problem. Otherwise the results from
+// processHTTPValidation are returned.
+func (va *ValidationAuthorityImpl) fetchHTTPSimple(
+	ctx context.Context,
+	host string,
+	path string) ([]byte, []core.ValidationRecord, *probs.ProblemDetails) {
+	body, records, err := va.processHTTPValidation(ctx, host, path)
+	if err != nil {
+		// Use detailedError to convert the error into a problem
+		return body, records, detailedError(err)
+	}
+	return body, records, nil
+}
+
+// fallbackErr returns true only for:
+//   1. net.OpError instances
+//   2. net.Error instances
+//   3. url.Error instances wrapping one of the above
+// fallbackErr returns false for all other errors. By policy we're only
+// interested in retrying requests that were made to an IPv6 address that
+// generated a network error.
+func fallbackErr(err error) bool {
+	// Err shouldn't ever be nil if we're considering it for fallback
+	if err == nil {
+		return false
+	}
+
+	// if the error is a URL error, unwrap it and see if the inner err is
+	// a fallback error
+	if urlErr, ok := err.(*url.Error); ok {
+		return fallbackErr(urlErr.Err)
+	}
+
+	// Network operation errors should provoke a fallback
+	if _, ok := err.(*net.OpError); ok {
```

Comment:
I think we additionally want to check for `err.Op == "dial"` so we don't catch timeouts that occur after a successful dial has already happened (for instance, timeouts during read of the HTTP response). There's an example of this in `detailedError`.

Similarly, I think the "Generic network errors" clause below may catch some cases that we don't want. I believe a failure during the dial will always be a `net.OpError` with `err.Op == "dial"`. Did you run across other cases in your testing?

Reply:
I think you're right. Looking carefully the Op is always "dial" for the cases I was able to invoke that didn't involve breaking something at a layer higher than TCP/IP.

Reply:
085fae8 removes the generic network error fallback and explicitly requires net error's have a "dial" op to qualify for fallback.